synapse/handlers/message.py

```diff
@@ -979,9 +980,43 @@ async def handle_new_client_event(
             logger.exception("Failed to encode content: %r", event.content)
             raise
 
-        await self.action_generator.handle_push_actions_for_event(event, context)
+        # We now persist the event (and update the cache in parallel, since we
+        # don't want to block on it).
+        result = await make_deferred_yieldable(
+            defer.gatherResults(
+                [
+                    run_in_background(
+                        self._persist_event,
+                        requester=requester,
+                        event=event,
+                        context=context,
+                        ratelimit=ratelimit,
+                        extra_users=extra_users,
+                    ),
+                    run_in_background(
+                        self.cache_joined_hosts_for_event, event, context
+                    ),
```

Comment:
if this fails do we want to cancel `_persist_event` (and swallow the result if it completes successfully)? Maybe give it a `.errback` which just logs the error?